### PR TITLE
applied Phabricator after cloning

### LINKED ISSUE #1268: Make the commit classifier script apply the Phabricator patch as its first step
Most failures happen when applying patches, so that should be the first step (right after cloning). This way, we don't waste time doing other things if the patch fails to apply.

### PR DESCRIPTION
**Pull request by Emeka-Onwuepe**
fixes  #1268

### CODE REVIEW

**Comment by Emeka-Onwuepe** [comment]
Hello sir, please kindly review this. Let me know if I am on the right track.

File: .gitignore [hunk 1]
```diff
@@ -17,6 +17,7 @@ http_service/models/*model*
 
 data/
 sheets/
+git_repo-shared/
```
Comment (suhaibmujahid) [inline .gitignore:20]: Unrelated change!

File: scripts/commit_classifier.py [hunk 2]
```diff
@@ -135,6 +135,8 @@ def __init__(
         method_defect_predictor_dir: str,
         use_single_process: bool,
         skip_feature_importance: bool,
+        phabricator_deployment=None,
+        diff_id=None,
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:139]: Could you please add the type hints?

File: scripts/commit_classifier.py [hunk 3]
```diff
@@ -597,10 +606,8 @@ def classify(
         self.update_commit_db()
 
         if phabricator_deployment is not None and diff_id is not None:
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:608]: Now we could check if `self.revision` has a value instead of this. Also, we could drop `phabricator_deployment` and `diff_id` from the arguments of the `classify()` method since they are not needed anymore (the assertions could be moved to the constructor).

File: scripts/commit_classifier.py
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:1]: Did you run the script to make sure the it is still run as expected?

**Review by suhaibmujahid — COMMENTED** [review]
Thank you, @Emeka-Onwuepe! Please see my comments.

File: scripts/commit_classifier.py
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:1]: I only tested with revision;

scripts.commit_classifier regressor repo_dir --revision REVISION 

which I believe will not trigger self.apply_phab().

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:1]: Can I get diff-id?

File: scripts/commit_classifier.py (hunk 3 above)
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:608]: Well noted

File: scripts/commit_classifier.py (hunk 2 above)
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:139]: Ok sir

File: .gitignore (hunk 1 above)
Comment (Emeka-Onwuepe) [reply .gitignore:20]: The folder was created when I ran 

scripts.commit_classifier regressor repo_dir --revision REVISION 

I will remove it

File: scripts/commit_classifier.py
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:1]: Tested with --diff-id=234724

File: scripts/commit_classifier.py [hunk 4]
```diff
@@ -582,25 +592,12 @@ def feature_sort_key(f):
...
-        diff_id: Optional[int] = None,
         runnable_jobs_path: Optional[str] = None,
     ) -> None:
-        if revision is not None:
-            assert phabricator_deployment is None
-            assert diff_id is None
-
-        if diff_id is not None:
-            assert phabricator_deployment is not None
-            assert revision is None
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:595]: The assertions here got lost!

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:595]: Phabricator_deployment and diff_id were no longer present in the method, so I didn't see the need to check for them.

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:595]: The assertions are trying to avoid the presence of revision, diff_id, and phabricator_delopment while ensuring that diff_id is provided whenever phabricator_development is provided with revision. 

Since both phabricator_development and diff_id are absent, I think the best place to put the assertions will be in the main function.

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:595]: I will fix it

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:595]: I have resolved it somewhere else, please have a second look

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:595]: I moved the assertions to main()

File: scripts/commit_classifier.py [hunk 5]
```diff
@@ -582,25 +592,12 @@ def feature_sort_key(f):
...
-
         self.update_commit_db()
 
-        if phabricator_deployment is not None and diff_id is not None:
-            with hglib.open(self.repo_dir) as hg:
-                self.apply_phab(hg, phabricator_deployment, diff_id)
-
-                revision = hg.log(revrange="not public()")[0].node.decode("utf-8")
+        if self.revision is not None:
+            revision = self.revision
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:600]: I would add `assert revision is None`  here in case the method gets called outside the main with the wrong args.

File: scripts/commit_classifier.py (hunk 4 above)
Comment (suhaibmujahid) [reply scripts/commit_classifier.py:595]: Thank you!

File: scripts/commit_classifier.py (hunk 5 above)
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:600]: Ok sir, noted

File: scripts/commit_classifier.py [hunk 6]
```diff
@@ -148,6 +150,14 @@ def __init__(
                 "hg::https://hg.mozilla.org/mozilla-central", git_repo_dir
             )
 
+        self.revision = None
+        if phabricator_deployment is not None and diff_id is not None:
+            with hglib.open(self.repo_dir) as hg:
+                self.apply_phab(hg, phabricator_deployment, diff_id)
+
+                self.revision = hg.log(revrange="not public()")[0].node.decode("utf-8")
+                assert self.revision is not None
```
[the remark this answers is not in the document]
Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py:159]: Well noted sir

File: scripts/commit_classifier.py [hunk 7]
```diff
@@ -135,6 +135,8 @@ def __init__(
         method_defect_predictor_dir: str,
         use_single_process: bool,
         skip_feature_importance: bool,
+        phabricator_deployment: str,
+        diff_id: int,
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:139]: The type hints here don't match the original ones:
https://github.com/mozilla/bugbug/blob/7a5e635999995b0ae603cdc960dfcdd06135358c/scripts/commit_classifier.py#L585-L586

Comment (Emeka-Onwuepe) [reply scripts/commit_classifier.py]: well noted, Sir

File: scripts/commit_classifier.py [hunk 8]
```diff
@@ -148,6 +150,13 @@ def __init__(
                 "hg::https://hg.mozilla.org/mozilla-central", git_repo_dir
             )
 
+        self.revision = None
+        if phabricator_deployment is not None and diff_id is not None:
```
Comment (suhaibmujahid) [inline scripts/commit_classifier.py:154]: nit: Checking `phabricator_deployment` as an assertion will help catch missing argument if the this get called directly in the future:
```suggestion
        if diff_id is not None:
            assert phabricator_deployment is not None
```

**Review by suhaibmujahid — APPROVED** [review]
Thank you, @Emeka-Onwuepe! LGTM.

**Review by suhaibmujahid — APPROVED** [review]

**Review by marco-c — APPROVED** [review]

**Comment by marco-c** [comment]
Thanks!
